_posts/2019-08-22-view-encapsulation-w-angularze.md

```diff
@@ -0,0 +1,445 @@
+
+Brak kapsułkowania, czyli style utworzone w komponencie są globalne (w sekcji `<head>`).  
+W tym trybie elementy HTML i odpowiadające im selektory CSS wyglądają tak samo jak te, które napisaliśmy w kodzie.  
+Może to spowodować niechciane nadpisywanie stylów lub dodawanie ich do elementów, które nie posiadają żadnego stylu.
+
+W przykładzie usunęliśmy styl paragrafu w komponencie `app-green`.  
+[Link do repozytorium](https://github.com/Michuu93/view-encapsulation-demo/tree/ViewEncapsulation.None)
+
+```typescript
+import {Component, ViewEncapsulation} from '@angular/core';
+
+@Component({
+    selector: 'app-red',
+    template: `
+        <p>Red paragraph!</p>
+    `,
+    styles: [`
+        p {
+            color: red;
+        }
+    `],
+    encapsulation: ViewEncapsulation.None
+})
+export class RedComponent {
+}
+```
+
+```typescript
+import {Component, ViewEncapsulation} from '@angular/core';
+
+@Component({
+    selector: 'app-green',
+    template: `
+        <p>Green paragraph!</p>
+    `,
+    encapsulation: ViewEncapsulation.None
+})
+export class GreenComponent {
+}
+```
+
+```typescript
+import {Component, ViewEncapsulation} from '@angular/core';
+
+@Component({
+    selector: 'app-blue',
+    template: `
+        <p>Blue paragraph!</p>
+    `,
+    styles: [`
+        p {
+            color: blue;
+        }
+    `],
+    encapsulation: ViewEncapsulation.None
+})
+export class BlueComponent {
+}
+```
+
+Wynikowy kod HTML:
+
+```html
+<head>
+    <style>
+        p {
+            color: red;
+        }
+    </style>
+    <style>
+        p {
+            color: blue;
+        }
+    </style>
+</head>
+
+<body>
+    <app-root ng-version="8.2.2">
+        <app-red>
+            <p>Red paragraph!</p>
+        </app-red>
+        <app-green>
+            <p>Green paragraph!</p>
+        </app-green>
+        <app-blue>
+            <p>Blue paragraph!</p>
+        </app-blue>
+    </app-root>
+</body>
+```
+
+![ViewEncapsulation.None](/assets/img/posts/2019-08-22-view-encapsulation/view_encapsulation_none_result.jpg)
+<span class="img-legend">Wynik widoczny w przeglądarce</span>
+
+Jak widzimy, style zostały dodane w sekcji `<head>`, co spowodowało nadpisanie pierwszego stylu paragrafu drugim - `color: blue`. W efekcie wszystkie paragrafy mają ten sam kolor, również paragraf z komponentu `app-green`, który nie posiada żadnego stylu i powinien mieć kolor domyślny.
+
+### ViewEncapsulation.Emulated (default)
+
+Domyślny tryb kapsułkowania w Angularze, w którym style są domknięte w komponencie.  
+W tym trybie style również znajdują się z sekcji `<head>`, ale posiadają dodatkowe atrybuty które wiążą je z elementami HTML pochodzącymi z tego samego komponentu.  
+Dzięki temu na stronie może istnieć kilka komponentów zawierających element tego samego typu, ale z różnymi stylami.  
+**Uwaga!** W tym trybie style rodzica nie mają wpływu na elementy dziecka (ponieważ każdy element otrzymuje własny, unikalny atrybut).
```

Comment:
Właśnie style rodzica mogą mieć wpływ na elementy potomne.

Jeżeli już chcemy pisać takie zdania "upraszczające" to lecimy z czymś takim:
Enakpsulowane w ten sposób elementy nie będą miały wpływu na inne elementy na stronie (ponieważ ich style są domknięte unikalnymi atrybutami elementów). Globalne style strony (oraz style innch komponentów) mogą mieć wpływ na element z emulowaną enakspulacją.


I wtedy w ShadoDom piszesz, że to już jest ultra level i tam, ani on nie wpływa na stornę, ani na niego nie wpływa strona.